org/mozilla/jss/util/jssutil.c

```diff
@@ -750,6 +752,106 @@ void JSS_DerefJString(JNIEnv *env, jstring str, const char *ref) {
+
+    // CERTIFICATE_ENCODING_EXCEPTION <-> SEC_ERROR_CERT_NOT_VALID
+    clazz = (*env)->FindClass(env, CERTIFICATE_ENCODING_EXCEPTION);
+    if (clazz != NULL && (*env)->IsInstanceOf(env, except, clazz)) {
+        PORT_SetError(SEC_ERROR_CERT_NOT_VALID);
+        return SECFailure;
+    }
+
+    // CERTIFICATE_EXPIRED_EXCEPTION <-> SEC_ERROR_EXPIRED_CERTIFICATE
+    clazz = (*env)->FindClass(env, CERTIFICATE_EXPIRED_EXCEPTION);
+    if (clazz != NULL && (*env)->IsInstanceOf(env, except, clazz)) {
+        PORT_SetError(SEC_ERROR_EXPIRED_CERTIFICATE);
+        return SECFailure;
+    }
+
+    // CERTIFICATE_NOT_YET_VALID_EXCEPTION <-> SEC_ERROR_CERT_NOT_VALID
+    clazz = (*env)->FindClass(env, CERTIFICATE_NOT_YET_VALID_EXCEPTION);
+    if (clazz != NULL && (*env)->IsInstanceOf(env, except, clazz)) {
+        PORT_SetError(SEC_ERROR_CERT_NOT_VALID);
+        return SECFailure;
+    }
+
+    // CERTIFICATE_PARSING_EXCEPTION <-> SEC_ERROR_BAD_DER
+    clazz = (*env)->FindClass(env, CERTIFICATE_PARSING_EXCEPTION);
+    if (clazz != NULL && (*env)->IsInstanceOf(env, except, clazz)) {
+        PORT_SetError(SEC_ERROR_BAD_DER);
+        return SECFailure;
+    }
+
+    // CERTIFICATE_REVOKED_EXCEPTION <-> SEC_ERROR_REVOKED_CERTIFICATE
+    clazz = (*env)->FindClass(env, CERTIFICATE_REVOKED_EXCEPTION);
+    if (clazz != NULL && (*env)->IsInstanceOf(env, except, clazz)) {
+        PORT_SetError(SEC_ERROR_REVOKED_CERTIFICATE);
+        return SECFailure;
+    }
+
+    return SECFailure;
+}
+
+/************************************************************************
+** JSS_SECStatusToException
+**
+** Convert a failing SECStatus and PRErrorCode combination into a raised
+** JNI exception.
+**
+** See jssutil.h for a list of supported exceptions.
+**
+*/
+void JSS_SECStatusToException(JNIEnv *env, SECStatus result, PRErrorCode code) {
+    JSS_SECStatusToExceptionMessage(env, result, code, "");
+}
+
+void JSS_SECStatusToExceptionMessage(JNIEnv *env, SECStatus result, PRErrorCode code, const char *message) {
+    if (result == SECSuccess) {
+        // We ignore PRErrorCode here. However, while we could clear an
+        // exception if it occurred, we don't; that's the caller's choice to
+        // make.
+        return;
+    }
+
+    if (result == SECFailure) {
+        switch (code) {
+            case SEC_ERROR_CERT_NOT_VALID:
+                JSS_throwMsgPrErrArg(env, CERTIFICATE_NOT_YET_VALID_EXCEPTION, message, code);
+                break;
+            case SEC_ERROR_EXPIRED_CERTIFICATE:
+                JSS_throwMsgPrErrArg(env, CERTIFICATE_EXPIRED_EXCEPTION, message, code);
+                break;
+            case SEC_ERROR_BAD_DER:
+                JSS_throwMsgPrErrArg(env, CERTIFICATE_PARSING_EXCEPTION, message, code);
+                break;
+            case SEC_ERROR_REVOKED_CERTIFICATE:
+                JSS_throwMsgPrErrArg(env, CERTIFICATE_REVOKED_EXCEPTION, message, code);
+                break;
+        }
```

Comment:
Do we need a default case and throw a generic exception?

Reply:
Added.